odl/trafos/wavelet.py

```diff
@@ -438,17 +467,42 @@ def __init__(self, range, wavelet, nlevels=None, pad_mode='constant',
         """
         super(WaveletTransformInverse, self).__init__(
             space=range, wavelet=wavelet, variant='inverse', nlevels=nlevels,
-            pad_mode=pad_mode, pad_const=pad_const, impl=impl)
+            pad_mode=pad_mode, pad_const=pad_const, impl=impl, axes=axes,
+            slices=slices, shapes=shapes)
 
     def _call(self, coeffs):
         """Return the inverse wavelet transform of ``coeffs``."""
+        if self._coeff_slices is None:
+            raise ValueError(
+                "self._coeff_slices not initialized.  Run the forward "
+                "transform once first to initialize it.")
         if self.impl == 'pywt':
-            shapes = pywt_coeff_shapes(self.range.shape, self.pywt_wavelet,
-                                       self.nlevels, self.pywt_pad_mode)
-            coeff_list = pywt_coeffs_from_flat_array(coeffs, shapes)
-            return pywt_multi_level_recon(
-                coeff_list, recon_shape=self.range.shape,
-                wavelet=self.pywt_wavelet, mode=self.pywt_pad_mode)
+            coeffs = pywt.unravel_coeffs(coeffs,
+                                         coeff_slices=self._coeff_slices,
+                                         coeff_shapes=self._coeff_shapes,
+                                         output_format='wavedecn')
+            recon = pywt.waverecn(
+                coeffs, wavelet=self.pywt_wavelet, mode=self.pywt_pad_mode,
+                axes=self.axes)
+            recon_shape = self.range.shape
+            if recon.shape != recon_shape:
```

Comment:
This could use some information describing what happens

Reply:
Sure. Due to the decimation/upsampling by two, it is possible for signals with odd shape to have have a shape rounded up to the next even size after reconstruction:

e.g. in 1D, downsampling by two of a length 7 or 8 signal gives a length 4 signal in each band. Upon reconstruction, upsampling by two results in a length 8 signal. The extra sample needs to be discarded in the case that the original signal was only length 7.

There same behavior was also in the old implementation:
https://github.com/odlgroup/odl/blob/b06a6a2540ce398184d4adbed964ac0bad315de9/odl/trafos/backends/pywt_bindings.py#L591-L607